### ISSUE TITLE
Diffraction did not work for gradient optimization

### ISSUE THREAD

**Issue by yuyuyuyao-ai** [NVlabs/sionna#1088]
I would like to maximize the received power at a fixed TX and RX location by tuning the rotation angle of a passive reflector.

The loss function is defined as the negative path gain computed by PathSolver. When only specular reflection is enabled, the loss function is discontinuous: as the reflector angle changes, a ray may either reach the RX or disappear entirely. This makes gradient-based optimization invalid. So I enabled diffraction and encountered the following error:

[Error_Log.docx](https://github.com/user-attachments/files/24457748/Error_Log.docx)

I tried detaching self._diffracting_wedges[0] in
.../sionna/rt/path_solvers/sample_data.py. This removes the error, but the kernel eventually crashes and needs to be restarted. Moreover, this modification seems to remove diffraction from gradient propagation altogether, so diffraction no longer influences the optimization.

Is this understanding correct? If so, are there recommended approaches for optimizing reflector orientation in this setting, or alternative loss formulations that avoid these issues?

Thanks a lot!

**Comment by ahmadfaisal9900** [comment]
hi, in Dr.Jit, arbitrary indexed writes into local/thread memory are currently not supported by reverse-mode AD. detaching self._diffracting_wedges[0] silences the error because you force that written value to become “non-AD”. but yes this also means no gradients can flow through.

i'd like to ask - have you looked into optimizing a surrogate, not discrete paths for this problem when your loss isn't continous?

a common trick may be discontinuity smoothing which entails replacing binary visibility with a smooth surrogate like a straight-through estimator or gumbel-softmax, then differentiate that surrogate.

you will end up needing to implement a custom solverpaths func though since Sionna’s PathSolver doesn’t currently expose this at the moment

**Comment by yuyuyuyao-ai** [comment]
Thanks for the information and suggestions!
